### PR TITLE
Authority control bugfixes

### PR DESCRIPTION
**Pull request by benbosman**
This contains general fixes to get the basics of authority control functional. It includes:
* Bugfixes to various parts of the authority control functionality
* Integration Tests
* Mock authority solr for the Integration Tests

This doesn't solve all bugs in authority control yet. The AuthorityRestRepositoryIT.correctCommonTypesTest() is currently disabled because it's not functional yet. This will require additional changes.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 1]
```diff
@@ -364,13 +365,13 @@ public ResourceSupport findRel(HttpServletRequest request, HttpServletResponse r
...
     @RequestMapping(method = RequestMethod.GET, value = REGEX_REQUESTMAPPING_IDENTIFIER_AS_STRING_VERSION_STRONG +
-        "/{rel}/{relid:[\\w+\\-]+}")
+        "/{rel}/{relid}")
     public ResourceSupport findRel(HttpServletRequest request, HttpServletResponse response,
                                    @PathVariable String apiCategory,
                                    @PathVariable String model, @PathVariable String id, @PathVariable String rel,
                                    @PathVariable String relid,
                                    Pageable page, PagedResourcesAssembler assembler,
-                                   @RequestParam(required = false) String projection) {
+                                   @RequestParam(required = false) String projection) throws Throwable {
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java]: I guess `throws Throwable` was left here during early developments

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 2]
```diff
@@ -754,7 +755,7 @@ public ResourceSupport findRel(HttpServletRequest request, HttpServletResponse r
                                                                            String id, String rel, String relid,
                                                                            Pageable page,
                                                                            PagedResourcesAssembler assembler,
-                                                                           String projection) {
+                                                                           String projection) throws Throwable {
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java]: I guess `throws Throwable` was left here during early developments

**Review by abollini — COMMENTED** [review]
This are just my quick finding, I hope to make a try in the coming days

File: dspace-api/src/main/java/org/dspace/content/authority/service/ChoiceAuthorityService.java [hunk 3]
```diff
@@ -168,4 +168,9 @@ public Choices getBestMatch(String fieldKey, String query, Collection collection
 
     public ChoiceAuthority getChoiceAuthorityByAuthorityName(String authorityName);
 
+    /**
+     * This method has been created to have a way of clearing the cache kept inside the service
+     */
+    public void clearController();
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/authority/service/ChoiceAuthorityService.java]: A minor thing, but I'd recommend renaming this to be `clearCache();` as that aligns better with the method description here.  The word "Controller" is confusing to me, as this is not a Spring Controller (which is the first thing that comes to mind here).

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        configurationService.setProperty("choices.presentation.dc.contributor.author",
+                "authorLookup");
+        configurationService.setProperty("authority.controlled.dc.contributor.author",
+                "true");
+
+        configurationService.setProperty("authority.author.indexer.field.1",
+                "dc.contributor.author");
+
+
+        //TODO Comments
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java]: Could we add in the comments here that are still TODO?

**Review by tdonohue — APPROVED** [review]
👍 Overall, this looks good to me, and it seems to have ITs to prove it is working.  I can admit however that I'm not as familiar with this area of the code, so not every change is 100% clear, but the ITs look correct.

That said, I did add a few minor inline comments.  Overall though, I approve of this.

**Review by abollini — COMMENTED** [review]
Hi @benbosman thanks for the PR and to include some initial test for the authority endpoint. Many fixes are related to the SOLRAuthority that we are not using to much. Can you clarify the changes in the solr schema? which issue they solve, do you have a test for it? it will be nice to add IT for use case that you discover to be fault so that this will guarantee us against future regression

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        getClient(token).perform(
+                get("/api/integration/authorities/SolrAuthorAuthority/entryValues/" + id))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.page.totalElements", Matchers.is(1)));
+    }
+
+    @Override
+    public void destroy() throws Exception {
+        AuthorityServiceFactory.getInstance().getAuthorityIndexingService().cleanIndex();
+        configurationService.setProperty("plugin.named.org.dspace.content.authority.ChoiceAuthority", null);
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java]: This entire `destroy()` method can actually be *removed*.  I had overlooked that all this is doing is resetting the ConfigurationService back to defaults.  That is *already performed* in `AbstractIntegrationTestWithDatabase.java` in the `destroy()` method...and all ITs inherit that behavior already.  See this line of code: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/test/java/org/dspace/app/rest/test/AbstractIntegrationTestWithDatabase.java#L187

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+    @Autowired
+    private PluginService pluginService;
+
+    @Autowired
+    private ChoiceAuthorityService cas;
+
+    @Before
+    public void setup() throws Exception {
+        super.setUp();
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java]: This isn't a worry. As noted below, the default test environment configuration gets automatically reset after each test.  We also already have a number of other ITs that modify the configuration via the API in a similar manner. See the `destroy` method here: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/test/java/org/dspace/app/rest/test/AbstractIntegrationTestWithDatabase.java#L187

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java]: thanks @tdonohue I haven't noted the reloadConfiguration in the abstract class

**Review by abollini — APPROVED** [review]
the PR is almost ready to merge, there are only minor comments inline to be addressed

File: dspace-api/src/main/java/org/dspace/content/authority/service/ChoiceAuthorityService.java (hunk 3 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/authority/service/ChoiceAuthorityService.java]: This has been renamed to clearCache

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java (hunk 1 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java]: This is explained at https://github.com/atmire/DSpace/blob/authority-control-bugfixes/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L779

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java (hunk 2 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java]: This is explained at https://github.com/atmire/DSpace/blob/authority-control-bugfixes/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L779

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorityRestRepositoryIT.java]: The ConfigurationService reset has been removed from the destroy() method.
The cleanIndex was still necessary, so this has been left in this section

**Pull request merged**